### PR TITLE
[DNM] arch review, please: sharding workflows to Cromwell v2

### PR DESCRIPTION
**Pull request by davidangb**
do not merge.

Looking for high-level architectural feedback on this approach to sharding workflows to Cromwell. Please don't spend too much time on syntax, as this may get refactored.

Functionally, I want to have an algorithm to route batches of workflows to a cromwell instance. Once submitted, I'd write the cromwell instance into the workflow's db row. Anything that queries cromwell about that workflow would read the cromwell instance from the Workflow or WorkflowRecord object, and route to that instance.

The basics of my approach in code are:
1. In Boot.scala, create an ExecutionServiceCluster object containing all the cromwell instances from config. ExecutionServiceCluster facades access to the ExecutionServiceDAOs representing those instances.
2. Anywhere we previously passed an ExecutionServiceDAO into the constructor for an Actor or monitor or whatever, we now pass an ExecutionServiceCluster
3. Anywhere we previously called a method on ExecutionServiceDAO, we now call that method on ExecutionServiceCluster, passing it the workflow(s) in question. ExecutionServiceCluster inspects the workflow(s), finds the right cromwell instance, and calls the same method on the right ExecutionServiceDAO.

In practice, I implemented ExecutionServiceCluster using polymorphic methods, to ease consumption. Is this overkill?

Still a whole lot left to do:
- actually implement the sharding algorithm, including persisting the cromwell instance onto a workflow once it's been submitted
- add a spate of tests
- have a more graceful way to create ExecutionServiceCluster in tests; the test code changes are messy.
- [ ] **Submitter**: Rebase to develop. DO NOT SQUASH
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: Make sure documentation for code is complete
- [ ] **Submitter**: Make sure liquibase is updated if appropriate
- [ ] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [ ] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
- [ ] **LR**: Initial review by LR and others.
- [ ] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [ ]  sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by helgridly** [comment]
This looks like it does some of the work I was intending to cover in my PR as a result of talking to Doug, so we should probably sync up on this.

I don't like the pattern of "ask the cluster for the appropriate DAO and then call the relevant function on the DAO". Just send the call to submit, status, whatever to the cluster and have it act as a facade. It can return both the workflow ID and some identifier for "which cromwell" when submitting, and the caller can be responsible for saving that "which cromwell" to the workflow row.

**Comment by davidangb** [comment]
Oooooh - good call @helgridly on using the cluster as a facade. I like that much better. I should probably just make ExecutionServiceCluster extend ExecutionServiceDAO and then the code changes can be much smaller.

**Comment by hjfbynara** [comment]
Trying to keep my comments limited to Rawls and not cromwell (and all that is involved with managing/running/configuring multiple cromwell instances).

The list of cromwell servers.  Is this a list of cromwell URLs?  DNS names? internal names?  Where/how does the mapping between the name that Rawls uses for each cromwell instance AND the actual instance (IP, URL,...) happen and how?

Do we need to care as part of this story - how Rawls deals when one or more of the cromwell instances is unavailable?  Do each have their own "queue" or is there one big queue that is used to feed the individual cromwells?  Today if I understand things it is somewhat simple - with one cromwell If it is down or unavailable queue new submissions and I guess report issue on status.  Adding more than one - brings up what to do in the various different possible scenarios.

  -Henry

**Comment by davidangb** [comment]
@hjfbynara the list of cromwell servers is being defined in another story (GAWB-778), though obviously there's a ton of overlap with this one. Maybe @helgridly can comment on that.

As far as rawls is concerned, there's a single queue. As we pull items off that queue, we choose (in batches) a cromwell to submit them to. Once they're submitted, we remember where they're running and will ask the right cromwell about them.

We definitely need to be concerned with an instance being unavailable ... as I hack through the code I will consider what complications sharding introduces over our current situation w/r/t a downed cromwell.

**Comment by helgridly** [comment]
Also, from my own thoughts plus talking to Doug, I think the config should probably look something like this:

```
executionservice : {
    servers : {
        "cromwell1" : {
            "path" : "cromwell1.broad.dsde-dev.firecloud.theinternet:1337",
            "readOnly" : true
        },
        "cromwell2" : {
            "path" : "cromwell2.broad.dsde-dev.firecloud.theinternet:4242",
            "readOnly" : false
        }
    }
}
```

**Comment by dvoet** [comment]
@helgridly why is there a readOnly attribute? Why would the form of path change from what we have now (e.g. https://cromwell.dsde-dev.broadinstitute.org:443/api)?

@hjfbynara I was not expecting how we address a particular cromwell to change from how it works today only we would have more than 1 URL.

**Comment by helgridly** [comment]
@dvoet There was discussion of "what if we want to decommission a Cromwell", which would require being able to query a Cromwell for currently-running workflows but not submit new ones to it.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ExecutionServiceCluster.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  override val errorReportSource = "cromwell"
+
+  // ====================
+  // facade methods
+  // ====================
+  def submitWorkflow(wdl: String, inputs: String, options: Option[String], userInfo: UserInfo): Future[ExecutionServiceStatus]
+
+  def submitWorkflows[T](wdl: String, inputs: Seq[String], options: Option[String], userInfo: UserInfo)(wf: T): Future[Seq[Either[ExecutionServiceStatus, ExecutionServiceFailure]]]
+
+  def status[T](id: String, userInfo: UserInfo)(wf: T): Future[ExecutionServiceStatus]
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ExecutionServiceCluster.scala]: I would have expected `status(id: String, execServiceId: String, userInfo: UserInfo)` here. Why isn't it that? We're going to have to add some notion of `execServiceId` to the workflow anyway; it seems reasonable to me that if the caller has to provide a workflow ID, they should also provide which cromwell it ran on.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ExecutionServiceCluster.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  * Created by davidan on 6/14/16.
+  */
+// TODO: keys for the map of members should probably be strings
+trait ExecutionServiceCluster extends ErrorReportable {
+  override val errorReportSource = "cromwell"
+
+  // ====================
+  // facade methods
+  // ====================
+  def submitWorkflow(wdl: String, inputs: String, options: Option[String], userInfo: UserInfo): Future[ExecutionServiceStatus]
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ExecutionServiceCluster.scala:19]: Needs to return "which cromwell this got dispatched to" as well.

**Comment by jmthibault79** [comment]
What are the pros/cons for having the caller of executionServiceCluster keep track of the Cromwell IDs per workflow instead of executionServiceCluster itself?

My intuition is to have the ExecutionServiceCluster trait _exactly_ match the existing ExecutionServiceDAO, and the ExecutionServiceCluster would then dispatch requests appropriately.

**Comment by helgridly** [comment]
@jmthibault79 pros: they're already keeping track of the workflow id (and the workflow table is clearly where it belongs anyway). otherwise the exec cluster would potentially have to re-lookup the workflow in order to get the cromwell id.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  // by nature, this is only called for workflows that have not yet been submitted.
+  // therefore, we want to send the workflows to the cromwell instance chosen
+  // by the routing algorithm defined in this class.
+  def submitWorkflows[T](wdl: String, inputs: Seq[String], options: Option[String], userInfo: UserInfo)(wf: T): Future[Seq[Either[ExecutionServiceStatus, ExecutionServiceFailure]]] =
+    nextAvailableMember(wf).submitWorkflows(wdl, inputs, options, userInfo)
+
+  // following are called on a workflow that has already been submitted.
+  // therefore, we want to use the cromwell instance that has been persisted
+  // onto that workflow.
+  def status[T](id: String, userInfo: UserInfo)(wf: T): Future[ExecutionServiceStatus] =
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: Why do we need the `wf` here?  We already have the id.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  // for an already-submitted workflow, get the instance to which it was submitted
+  private def getMember[T](obj: T):ExecutionServiceDAO = obj match {
+      case wf:Workflow => getMember(wf)
+      case wr:WorkflowRecord => getMember(wr)
+      case seq:Seq[WorkflowRecord] => getMember(seq)
+      case _ => throw new RawlsException("Must be a Workflow, WorkflowRecord, or Seq[WorkflowRecord]")
+    }
+
+  // for unsubmitted workflows, get the best instance to which we should submit
+  private def nextAvailableMember[T](obj: T):ExecutionServiceDAO = obj match {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: If it's the best available member, it should be named accordingly.  The implementation of "best" might very well be "next" but we should separate the two concepts to clarify intention.

**Comment by davidangb** [comment]
@helgridly @jmthibault79 I am currently asking the caller to pass an entire Workflow or WorkflowRecord, and handling all routing within the Cluster class, because:
- for the read methods - status, outputs, logs, etc - we need more than just the workflow id; we need the execServiceId or whatever gets persisted. And I don't want to query the db again just to get that.
- I want to be flexible about the implementation, both for routing submissions and for querying. If we move from pure sharding to something that supports call-caching, for instance, we may need to inspect more properties of the workflows in order to route.

Whether I pass in a full Workflow/WorkflowRecord or just an execServiceId, I'd need to make updates to the ExecutionServiceDAO signature to allow ExecutionServiceCluster to match ExecutionServiceDAO. I wasn't ready to do that.

@helgridly totally right - the submit\* methods need to return a cromwell id. Since I haven't implemented the persistence yet, I just missed that.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala (hunk 3 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: the sequence of events is going to look like this:
- call submitWorkflows, pass a batch of workflows as an argument
- route those workflows to the best cromwell instance
- persist, on each of those workflows, the cromwell instance to which it was submitted
- when calling status, pass in the Workflow/WorkflowRecord
- status() reads the Workflow/WorkflowRecords, extracts the cromwell instance that was persisted
- cluster.status() calls instance.status() on the right instance

tl;dr we need more than the id to get the right instance.

though - we could eliminate id, if we have wf.

**Comment by helgridly** [comment]
assuming doug is right and we don't care about readOnly, the config will look like:

```
executionservice : {
    servers : {
        "cromwell1" : "cromwell1.broad.dsde-dev.firecloud.theinternet:1337",
        "cromwell2" : "cromwell2.broad.dsde-dev.firecloud.theinternet:4242"
    }
}
```

which means that the config-y code in boot.scala to extract it is:

```
val servers = executionServiceConfig.getConfig("servers")
servers.entrySet.map({ mapEntry =>
  mapEntry.getKey -> servers.getString(mapEntry.getKey)
}).toMap
```

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ExecutionServiceCluster.scala [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  def status[T](id: String, userInfo: UserInfo)(wf: T): Future[ExecutionServiceStatus]
+
+  def callLevelMetadata[T](id: String, userInfo: UserInfo)(wf: T): Future[ExecutionMetadata]
+
+  def outputs[T](id: String, userInfo: UserInfo)(wf: T): Future[ExecutionServiceOutputs]
+
+  def logs[T](id: String, userInfo: UserInfo)(wf: T): Future[ExecutionServiceLogs]
+
+  def abort[T](id: String, userInfo: UserInfo)(wf: T): Future[Try[ExecutionServiceStatus]]
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ExecutionServiceCluster.scala]: I think I may be piling on to things I have heard discussed, but I don't really like the (wf: T) stuff. Maybe it would be better with some super class that both Workflow and WorkflowRecord extended from but the way it stands it too untyped. I think these functions should take a WorkflowRecord instead of id.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+  // ====================
+  // facade-to-cluster entry points
+  // ====================
+  // for an already-submitted workflow, get the instance to which it was submitted
+  private def getMember(wfe: WorkflowExecution):Future[ExecutionServiceDAO] = wfe.executionServiceId match {
+    case Some(execId) => Future(getMember(execId))
+    case None => {
+      dataSource.inTransaction { dataAccess =>
+        dataAccess.workflowQuery.getExecutionServiceKey(wfe.id) map {execIdOption =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: why wouldn't we already have the key? I am concerned about nested transactions and a query per workflow model.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  def abort(wfe: WorkflowExecution, userInfo: UserInfo): Future[Try[ExecutionServiceStatus]] =
+    getMember(wfe).flatMap {memb => memb.abort(wfe.id, userInfo)}
+
+
+  // ====================
+  // facade-to-cluster entry points
+  // ====================
+  // for an already-submitted workflow, get the instance to which it was submitted
+  private def getMember(wfe: WorkflowExecution):Future[ExecutionServiceDAO] = wfe.executionServiceId match {
+    case Some(execId) => Future(getMember(execId))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: use Future.successful: Future.successful(getMember(execId))

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    memberArray(shardIndex)
+  }
+
+
+
+  // ====================
+  // clustering methods
+  // ====================
+  private def getMember(strKey: String): ExecutionServiceDAO = getMember(ExecutionServiceId(strKey))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: doesn't having the convenience method defeat the purpose of having a ExecutionServiceId class? Why not just have the key be a string?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala [hunk 9]
```diff
@@ -244,6 +247,19 @@ trait WorkflowComponent {
...
 
+    def batchUpdateStatusAndExecutionServiceKey(workflows: Seq[WorkflowRecord], newStatus: WorkflowStatus, execServiceId: ExecutionServiceId): ReadWriteAction[Int] = {
+      if (workflows.isEmpty) {
+        DBIO.successful(0)
+      } else {
+        UpdateWorkflowStatuAndExecutionIdRawSql.actionForWorkflowRecs(workflows, newStatus, execServiceId) flatMap { rows =>
+          if (rows.head == workflows.size)
+            DBIO.successful(workflows.size)
+          else
+            throw new RawlsConcurrentModificationException(s"could not update ${workflows.size - rows.head} workflows because their record version(s) have changed")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala]: this should be DBIO.failed(exception) instead of throw new exception but really you don't need the flatMap. Use map instead:
map { rows => if (rows.head == workflows.size) workflows.size else throw new exception }

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala (hunk 6 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala]: We have the executionServiceKey for WorkflowRecord objects, but not for Workflow objects. We could:
1. Update Workflow to include the executionServiceKey - but that means it will be visible to end users (at least end users of the API)
   - maybe it's possible to write a custom json serializer for Workflow, so we never send executionServiceKey out via the API, but we have it internally to the code
2. Update anywhere that calls this to use WorkflowRecord objects, not Workflow objects. This is pretty pervasive - SubmissionMonitorActor.abortActiveWorkflows(), WorkspaceService.deleteWorkspace(), WorkspaceService.workflowOutputs(), for instance.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala (hunk 9 above)
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala]: Hmm ... good call. I just copied this from batchUpdateStatus() above it. Should I update both?

**Comment by davidangb** [comment]
Closing in order to start a clean PR
